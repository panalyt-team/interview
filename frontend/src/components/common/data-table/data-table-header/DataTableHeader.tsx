import { makeStyles } from "@material-ui/styles";
import React, { useState } from "react";
import FilterListIcon from '@material-ui/icons/FilterList';
import { CheckboxGroupProperties } from "../../interfaces";
import { IconButton, Popover } from "@material-ui/core";
import CheckboxGroup from "../../checkbox-group/CheckboxGroup";
import createBreakpoints from "@material-ui/core/styles/createBreakpoints";

export interface DataTableHeaderProps {
    columns: string[];
    filters: CheckboxGroupProperties;
}

const breakpoints = createBreakpoints({});

const useStyles = makeStyles({
    root: {
        display: "grid",
        gridTemplateColumns: "repeat(auto-fit, minmax(100px, 1fr))",
        position: "relative",
        textAlign: "center",
        width: "100%",
        height: "70px",
        justifyContent: "space-around",
        alignItems: "center",
        backgroundColor: "#04a599",
        color: "#fafafa",
    },
    label: {
        fontSize: "1.2em",
        letterSpacing: "0.03em"
    },
    filterButton: {
        position: "absolute",
        right: "20px",
        [breakpoints.down("xs")]: {
            right: 0
        }
    },
    popoverContent: {
        padding: "10px"
    }
}, { classNamePrefix: "data-table-header" });

export const DataTableHeader = ({ columns, filters }: DataTableHeaderProps): JSX.Element => {
    const classes = useStyles();
    const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

    const handleFilterClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
        setAnchorEl(event.currentTarget);
    };

    const handlePopoverClose = (): void => {
        setAnchorEl(null);
    };

    const isPopoverOpened = Boolean(anchorEl);
    const id = isPopoverOpened ? 'simple-popover' : undefined;

    return (
        <div className={classes.root} data-test-id="header">
            {
                columns.map((column: string, index: number) =>
                    <p key={index} className={classes.label} title={column} data-test-id="column">{column}</p>
                )
            }

            <IconButton className={classes.filterButton} onClick={handleFilterClick} data-test-id="filters-button">
                <FilterListIcon />
            </IconButton>

            <Popover
                id={id}
                open={isPopoverOpened}
                anchorEl={anchorEl}
                onClose={handlePopoverClose}
                anchorOrigin={{
                    vertical: 'bottom',
                    horizontal: 'center',
                }}
                transformOrigin={{
                    vertical: 'top',
                    horizontal: 'center',
                }}>
                <div className={classes.popoverContent}>
                    <CheckboxGroup groupLabel={filters.groupLabel}
                        checkboxes={filters.checkboxes}
                        changeHandler={filters.changeHandler}
                    />
                </div>
            </Popover>

        </div>
    )
};

export default DataTableHeader;