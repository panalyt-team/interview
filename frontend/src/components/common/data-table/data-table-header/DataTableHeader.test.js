import { shallow } from 'enzyme';
import DataTableHeader from './DataTableHeader';

describe("DataTableHeader", () => {

    let filters = [{ "checked": true, "label": "Japan" }, { "checked": false, "label": "China" }, { "checked": false, "label": "Malaysia" }];

    it("renders correctly", () => {
        const wrapper = shallow(<DataTableHeader columns={["col1", "col2", "col3"]} filters={filters} />);

        expect(wrapper).toMatchSnapshot();
    });

    it("renders a correct amount of columns", () => {
        const wrapper = shallow(<DataTableHeader columns={["col1", "col2", "col3"]} filters={filters} />);

        const columns = wrapper.find('p[data-test-id="column"]');

        expect(columns.length).toBe(3);
    });

    it("opens a popover with filters when filters button is clicked", () => {
        const wrapper = shallow(<DataTableHeader columns={["col1", "col2", "col3"]} filters={filters} />);

        const buttonElement = wrapper.find('[data-test-id="filters-button"]');
        buttonElement.simulate('click', { currentTarget: buttonElement });
        const popoverElement = wrapper.find("#simple-popover");

        expect(popoverElement.exists()).toBeTruthy();
    });

});