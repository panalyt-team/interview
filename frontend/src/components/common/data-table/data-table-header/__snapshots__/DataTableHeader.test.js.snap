// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableHeader renders correctly 1`] = `ShallowWrapper {}`;
